src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts

```diff
@@ -410,36 +423,13 @@ export class ObjectExplorerService implements IObjectExplorerService {
 						allProviders.push(...nodeProviders);
 					}
 
-					self._sessions[session.sessionId!].nodes[node.nodePath].expandEmitter.event((expandResult: NodeExpandInfoWithProviderId) => {
-						if (expandResult && expandResult.providerId) {
-							resultMap.set(expandResult.providerId, expandResult);
-							// If we got an error result back then send error our error event
-							// We only do this for the MSSQL provider
-							if (expandResult.errorMessage && expandResult.providerId === mssqlProviderName) {
-								const errorType = expandResult.errorMessage.indexOf('Object Explorer task didn\'t complete') !== -1 ? 'Timeout' : 'Other';
-								// For folders send the actual name of the folder (since the nodeTypeId isn't useful in this case and the names are controlled by us)
-								const nodeType = node.nodeTypeId === NodeType.Folder ? node.label : node.nodeTypeId;
-								this._telemetryService.createErrorEvent(TelemetryKeys.TelemetryView.Shell, TelemetryKeys.TelemetryError.ObjectExplorerExpandError, undefined, errorType)
-									.withAdditionalProperties({
-										nodeType,
-										providerId: expandResult.providerId
-									}).send();
-							}
-
-						} else {
-							this.logService.error('OE provider returns empty result or providerId');
-						}
+					self._sessions[session.sessionId!].nodes[node.nodePath].expandEmitter.event(
+						(expandResult: NodeExpandInfoWithProviderId) =>
+							resolve(this.completedExpandRequest(expandResult, resultMap, node, newRequest, session, allProviders)));
```

Comment:
So this seems like it happens to work but it ends up leaving a bunch of dangling promises which never get handled. Each expandEmitter event is going to trigger this which will kick off a call to this.completedExpandRequest for each provider. But then only one will ever complete - the last one that "finishes" the map.

Reply:
That's current design I think

Reply:
No, because previously there was only one Promise (the outer one). Your refactor here added this new helper method which also returns a Promise called within the event handler itself that you aren't even handling.